metadata-ingestion/src/datahub/ingestion/source/feast_repository.py

```diff
@@ -0,0 +1,371 @@
+                )
+            )
+
+        if feature_view.stream_source is not None:
+            stream_source_platform, stream_source_name = self._get_data_source_details(
+                feature_view.stream_source
+            )
+            sources.append(
+                builder.make_dataset_urn(
+                    stream_source_platform,
+                    f"{project}.{stream_source_name}",
+                    self.source_config.environment,
+                )
+            )
+
+        return sources
+
+    def _get_entity(
+        self, project: str, feature_view: FeatureView, entity: Entity
+    ) -> MetadataWorkUnit:
+        """
+        Generate an MLPrimaryKey workunit for a Feast entity.
+        """
+
+        feature_view_name = f"{project}.{feature_view.name}"
+
+        entity_snapshot = MLPrimaryKeySnapshot(
+            urn=builder.make_ml_primary_key_urn(feature_view_name, entity.name),
+            aspects=[],
+        )
+
+        entity_snapshot.aspects.append(
+            MLPrimaryKeyPropertiesClass(
+                description=entity.description,
+                dataType=self._get_field_type(entity.value_type, entity.name),
+                sources=self._get_data_sources(project, feature_view),
+            )
+        )
+
+        mce = MetadataChangeEvent(proposedSnapshot=entity_snapshot)
+
+        return MetadataWorkUnit(id=entity.name, mce=mce)
+
+    def _get_feature(
+        self,
+        project: str,
+        feature_view: Union[FeatureView, OnDemandFeatureView],
+        feature: Feature,
+    ) -> MetadataWorkUnit:
+        """
+        Generate an MLFeature workunit for a Feast feature.
+        """
+
+        feature_view_name = f"{project}.{feature_view.name}"
+
+        feature_snapshot = MLFeatureSnapshot(
+            urn=builder.make_ml_feature_urn(feature_view_name, feature.name), aspects=[]
+        )
+
+        feature_sources = []
+
+        if isinstance(feature_view, FeatureView):
+            feature_sources = self._get_data_sources(project, feature_view)
+        elif isinstance(feature_view, OnDemandFeatureView):
+            if feature_view.input_request_data_sources is not None:
+                for (
+                    _,
+                    request_source,
+                ) in feature_view.input_request_data_sources.items():
+                    source_platform, source_name = self._get_data_source_details(
+                        request_source
+                    )
+
+                    feature_sources.append(
+                        builder.make_dataset_urn(
+                            source_platform,
+                            f"{project}.{source_name}",
+                            self.source_config.environment,
+                        )
+                    )
+
+            if feature_view.input_feature_views is not None:
+                for _, feature_view_source in feature_view.input_feature_views.items():
```

Comment:
If the key is unused here, could we use `.values()` instead?

Reply:
Thanks for finding it!